src/relay/transforms/annotate_target.cc

```diff
@@ -146,13 +179,19 @@ class AnnotateTargetRewriter : public ExprRewriter {
       // Bypass compiler begin due to lack of target information. It will be processed
       // when the following op handling arguments.
       ICHECK_EQ(pre->args.size(), 1U);
-      return post.as<CallNode>()->args[0];
+      // Preserve annotations
+      return post;
     } else if (op_node && pre->op == CompilerEndOp()) {
       // Override compiler end with the new target.
       ICHECK_EQ(pre->args.size(), 1U);
       auto input_expr = post.as<CallNode>()->args[0];
+      // Already annotated. Recover target
+      if (op_expr_to_target_.find(input_expr) == op_expr_to_target_.end()) {
+        op_expr_to_target_[input_expr] = post.as<CallNode>()->attrs.as<CompilerAttrs>()->compiler;
+      }
```

Comment:
Looks like you don't need the IF? Even `input_expr` is already in `op_expr_to_target_`, you can still override it, as suggested by the comment in L188. Accordingly, if you will override the target, you need `InsertAnnotation`.

Reply:
Removed IF